File: asm/src/main/java/dev/openrs2/asm/classpath/Library.java
package dev.openrs2.asm.classpath;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.SequenceInputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;
import java.util.jar.JarEntry;
import java.util.jar.JarInputStream;
import java.util.jar.JarOutputStream;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import dev.openrs2.asm.remap.ClassForNameRemapper;
import dev.openrs2.util.io.DeterministicJarOutputStream;
import dev.openrs2.util.io.SkipOutputStream;
import org.apache.harmony.pack200.Pack200;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.commons.ClassRemapper;
import org.objectweb.asm.commons.Remapper;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.util.CheckClassAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class Library implements Iterable<ClassNode> {
	private static final Logger logger = LoggerFactory.getLogger(Library.class);

	private static final String CLASS_SUFFIX = ".class";
	private static final String TEMP_PREFIX = "tmp";
	private static final String JAR_SUFFIX = ".jar";
	private static final byte[] GZIP_HEADER = { 0x1f, (byte) 0x8b };

	public static Library readJar(Path path) throws IOException {
		logger.info("Reading jar {}", path);

		var library = new Library();

		try (var in = new JarInputStream(Files.newInputStream(path))) {
			JarEntry entry;
			while ((entry = in.getNextJarEntry()) != null) {
				if (!entry.getName().endsWith(CLASS_SUFFIX)) {
					continue;
				}

				var clazz = new ClassNode();
				var reader = new ClassReader(in);
				reader.accept(clazz, ClassReader.SKIP_DEBUG);

				library.add(clazz);
			}
		}

		return library;
	}

	public static Library readPack(Path path) throws IOException {
		logger.info("Reading pack {}", path);

		var temp = Files.createTempFile(TEMP_PREFIX, JAR_SUFFIX);
		try {
			try (
				var header = new ByteArrayInputStream(GZIP_HEADER);
				var data = Files.newInputStream(path);
				var in = new GZIPInputStream(new SequenceInputStream(header, data));
				var out = new JarOutputStream(Files.newOutputStream(temp))
			) {
				Pack200.newUnpacker().unpack(in, out);
				return readJar(temp);
			}
		} finally {
			Files.deleteIfExists(temp);
		}
	}

	private final Map<String, ClassNode> classes = new TreeMap<>();

	public Library() {
		/* empty */
	}

	public Library(Library library) {
		for (var clazz : library.classes.values()) {
			var copy = new ClassNode();
			clazz.accept(copy);
			add(copy);
		}
	}

	public boolean contains(String name) {
		return classes.containsKey(name);
	}

	public ClassNode get(String name) {
		return classes.get(name);
	}

	public ClassNode add(ClassNode clazz) {
		return classes.put(clazz.name, clazz);
	}

	public ClassNode remove(String name) {
		return classes.remove(name);
	}

	@Override
	public Iterator<ClassNode> iterator() {
		return classes.values().iterator();
	}

	public void remap(Remapper remapper) {
		var classNames = new HashSet<String>();

		for (var clazz : classes.values()) {
			for (var method : clazz.methods) {
				if ((method.access & (Opcodes.ACC_NATIVE | Opcodes.ACC_ABSTRACT)) == 0) {
					ClassForNameRemapper.remap(remapper, method);
				}
			}

			classNames.add(clazz.name);
		}

		for (var name : classNames) {
			var in = classes.remove(name);

			var out = new ClassNode();
			in.accept(new ClassRemapper(out, remapper));

			classes.put(out.name, out);
		}
	}

	public void writeJar(Path path) throws IOException {
		logger.info("Writing jar {}", path);

		try (var out = new DeterministicJarOutputStream(Files.newOutputStream(path))) {
			for (var clazz : classes.values()) {
				var writer = new ClassWriter(0);

				clazz.accept(new CheckClassAdapter(writer, true));

				out.putNextEntry(new JarEntry(clazz.name + CLASS_SUFFIX));
				out.write(writer.toByteArray());
			}
		}
	}

	public void writePack(Path path) throws IOException {
		logger.info("Writing pack {}", path);

		var temp = Files.createTempFile(TEMP_PREFIX, JAR_SUFFIX);
		try {
			writeJar(temp);

			try (
				var in = new JarInputStream(Files.newInputStream(temp));
				var out = new GZIPOutputStream(new SkipOutputStream(Files.newOutputStream(path), GZIP_HEADER.length))
			) {
				Pack200.newPacker().pack(in, out);
			}
		} finally {
			Files.deleteIfExists(temp);
		}
	}
}

File: asm/src/main/java/dev/openrs2/asm/classpath/Library.kt
package dev.openrs2.asm.classpath

import com.github.michaelbull.logging.InlineLogger
import dev.openrs2.asm.io.LibraryReader
import dev.openrs2.asm.io.LibraryWriter
import dev.openrs2.util.io.useAtomicOutputStream
import org.objectweb.asm.tree.ClassNode
import java.nio.file.Files
import java.nio.file.Path
import java.util.SortedMap
import java.util.TreeMap

class Library(val name: String) : Iterable<ClassNode> {
    private var classes: SortedMap<String, ClassNode> = TreeMap()

    constructor(name: String, library: Library) : this(name) {
        for (clazz in library.classes.values) {
            val copy = ClassNode()
            clazz.accept(copy)
            add(copy)
        }
    }

    operator fun contains(name: String): Boolean {
        return classes.containsKey(name)
    }

    operator fun get(name: String): ClassNode? {
        return classes[name]
    }

    fun add(clazz: ClassNode): ClassNode? {
        return classes.put(clazz.name, clazz)
    }

    fun remove(name: String): ClassNode? {
        return classes.remove(name)
    }

    override fun iterator(): Iterator<ClassNode> {
        return classes.values.iterator()
    }

    fun remap(remapper: ExtendedRemapper) {
        classes = LibraryRemapper(remapper, classes).remap()
    }

    fun write(path: Path, writer: LibraryWriter, classPath: ClassPath) {
        logger.info { "Writing library $path" }

        path.useAtomicOutputStream { output ->
            writer.write(output, classPath, classes.values)
        }
    }

    companion object {
        private val logger = InlineLogger()

        fun read(name: String, path: Path, reader: LibraryReader): Library {
            logger.info { "Reading library $path" }

            val classes = Files.newInputStream(path).use { input ->
                reader.read(input)
            }

            val library = Library(name)
            for (clazz in classes) {
                library.add(clazz)
            }
            return library
        }
    }
}
